From 275571acc88f38a6d8856b620517943823bfacc3 Mon Sep 17 00:00:00 2001
From: Arnaud Pouliquen <arnaud.pouliquen@foss.st.com>
Date: Wed, 16 Jul 2025 11:10:49 +0200
Subject: [PATCH] remoteproc: stm32_rproc: Make MCU reset and hold_boot
 optional

On the STM32MP2x series, the Cortex-M33 boots first, and in this scenario,
Linux does not have the appropriate access to the reset controls.

In this configuration, the remoteproc driver attaches to the Cortex-M33
but should not manage the resets.

This patch makes the hold_boot and mcu reset declaration in the device
tree optional.

Change-Id: I01e06a2a1b39508688666de3df160de287ef85d0
Signed-off-by: Arnaud Pouliquen <arnaud.pouliquen@foss.st.com>
Reviewed-on: https://gerrit.st.com/c/mpu/oe/st/linux-stm32/+/475151
ACI: CITOOLS <MDG-smet-aci-reviews@list.st.com>
ACI: CIBUILD <MDG-smet-aci-builds@list.st.com>
Tested-by: Arnaud POULIQUEN <arnaud.pouliquen@st.com>
Domain-Review: Arnaud POULIQUEN <arnaud.pouliquen@st.com>
Reviewed-by: Arnaud POULIQUEN <arnaud.pouliquen@st.com>
Reviewed-by: Olivier MOYSAN <olivier.moysan@foss.st.com>
---
 drivers/remoteproc/stm32_rproc.c | 13 ++++---------
 1 file changed, 4 insertions(+), 9 deletions(-)

--- a/drivers/remoteproc/stm32_rproc.c
+++ b/drivers/remoteproc/stm32_rproc.c
@@ -668,7 +668,7 @@ static int stm32_rproc_set_hold_boot(str
 	struct stm32_rproc *ddata = rproc->priv;
 	struct stm32_syscon hold_boot = ddata->hold_boot;
 	struct arm_smccc_res smc_res;
-	int val, err;
+	int val, err = 0;
 
 	/*
 	 * Three ways to manage the hold boot
@@ -690,7 +690,7 @@ static int stm32_rproc_set_hold_boot(str
 		arm_smccc_smc(STM32_SMC_RCC, STM32_SMC_REG_WRITE,
 			      hold_boot.reg, val, 0, 0, 0, 0, &smc_res);
 		err = smc_res.a0;
-	} else {
+	} else if (hold_boot.map) {
 		/* Use syscon */
 		err = regmap_update_bits(hold_boot.map, hold_boot.reg,
 					 hold_boot.mask, val);
@@ -1071,7 +1071,7 @@ static int stm32_rproc_parse_dt(struct p
 		/* Try legacy fallback method: get it by index */
 		ddata->rst = devm_reset_control_get_by_index(dev, 0);
 	}
-	if (IS_ERR(ddata->rst))
+	if (IS_ERR(ddata->rst) && PTR_ERR(ddata->rst) != -ENOENT)
 		return dev_err_probe(dev, PTR_ERR(ddata->rst),
 				     "failed to get mcu_reset\n");
 
@@ -1106,12 +1106,7 @@ static int stm32_rproc_parse_dt(struct p
 
 	if (!ddata->hold_boot_rst && !ddata->hold_boot_smc) {
 		/* Default: hold boot manage it through the syscon controller */
-		err = stm32_rproc_get_syscon(np, "st,syscfg-holdboot",
-					     &ddata->hold_boot);
-		if (err) {
-			dev_err(dev, "failed to get hold boot\n");
-			return err;
-		}
+		stm32_rproc_get_syscon(np, "st,syscfg-holdboot",  &ddata->hold_boot);
 	}
 
 	err = stm32_rproc_get_syscon(np, "st,syscfg-pdds", &ddata->pdds);
